src/utils/locale.ts

```diff
@@ -150,14 +171,103 @@ export interface LangComponent extends GlobalAttrComponent {
    * @mdn [lang](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/lang)
    */
   locale?: string;
+
+  /**
+   * Used to store the effective locale to avoid multiple lookups.
+   *
+   * This prop should use the `@State` decorator and be initialized to "";
+   *
+   * Components should watch this prop to ensure messages are updated.
+   *
+   * @Watch("effectiveLocale")
+   * effectiveLocaleChange(): void {
+   *   updateMessages(this, this.effectiveLocale);
+   * }
+   */
+  effectiveLocale: string;
+}
+
+const connectedComponents = new Set<LocalizedComponent>();
+
+/**
+ * This utility sets up internals for messages support.
+ *
+ * It needs to be called in `connectedCallback`
+ *
+ * @param component
+ */
+export function connectLocalized(component: LocalizedComponent): void {
+  updateEffectiveLocale(component);
+
+  if (connectedComponents.size === 0) {
+    mutationObserver.observe(document.documentElement, {
+      attributes: true,
+      attributeFilter: ["lang"],
+      subtree: true
+    });
+  }
+
+  connectedComponents.add(component);
 }
 
+/**
+ * This is only exported for components that implemented the now deprecated `locale` prop.
+ *
+ * Do not use this utils for new components.
+ *
+ * @param component
+ */
+export function updateEffectiveLocale(component: LocalizedComponent): void {
+  component.effectiveLocale = getLocale(component);
+}
+
+/**
+ * This utility tears down internals for messages support.
+ *
+ * It needs to be called in `disconnectedCallback`
+ *
+ * @param component
+ */
+export function disconnectLocalized(component: LocalizedComponent): void {
+  connectedComponents.delete(component);
+
+  if (connectedComponents.size === 0) {
+    mutationObserver.disconnect();
+  }
+}
+
+const mutationObserver = createObserver("mutation", (records) => {
+  records.forEach((record) => {
+    const el = record.target as HTMLElement;
+
+    connectedComponents.forEach((component) => {
+      if (component.locale || (component.el.lang && el !== component.el) || !el.contains(component.el)) {
+        return;
+      }
+
+      const closestLangEl = closestElementCrossShadowBoundary<HTMLElement>(component.el, "[lang]");
```

Comment:
Great question. The mutation observer is configured to only run on the `lang` attribute, so I think we don't need to debounce yet. Local testing looks good, but we should really ask for additional testing from other teams to cover more complex scenarios. I'm thinking that for most use cases, it may not be an issue since locale wouldn't switch often. The only exception that comes to mind are builder apps, but even then, I imagine there'd only be one lang that changed and not multiple levels.

Reply:
Follow-up on the performance side of things:

* The util's mutation observer only observes mutations within light DOM
* We've updated the pattern so that composite widgets set the internal `effectiveLocale` prop, the assumption is that the top-most composite widget will handle lang and pass it along
* Changes to unrelated subtrees will be ignored